RFC-0030-Consolidate-Elastic-and-TorchX.md

```diff
@@ -0,0 +1,262 @@
+     2. Example: dealing with host failures in a job:
+        1. torchelastic alone cannot deal with host failures. The job needs to be configured (done via torchx) for
+           the scheduler to deal with host failures (e.g. replace the failed host).
+        2. Conversely, the scheduler replacing failed hosts for a job is not enough, the job itself needs to be authored
+           in such a way that surviving hosts can wait-for a failed host to be replaced then admit the newly replace host
+           into the worker group (provided by torchelastic).
+3. **Out-of-the-box support for job launchers in PyTorch** 
+    1. Without additional installs, PyTorch users can run PyTorch scripts both locally and remotely onto widely used schedulers 
+       (Kubernetes, SLURM, CloudProvider-specific batch schedulers).
+4. **Well-defined and self-service plugin-points to integrate PyTorch with the users' infrastructure.**
+    1. Examples include (but not limited to):
+       1. Job/Experiment tracking (e.g. MLflow, Weights&Biases, FBLearner, etc)
+       2. Registration of custom compute resource types to launch jobs onto (e.g. p4d.24xlarge [AWS], a2-megagpu-16g [GCP])
+       3. Code containerization (e.g. Docker, XAR, etc)
+ 
+
+## **Proposed Implementation**
+
+The diagram below depicts the proposed module move/upstream (optional merge of the `multiprocessing` module not shown).
+
+![`torch.x` Before and After](RFC-0030-assets/modules_before_after.png)
+
+**The following changes are proposed (refer to diagram above):**
+
+1. Move `torch.distribted.elastic.*` &rarr; `torch.x.elastic.*`
+2. Upstream `torchx.*` &rarr; `torch.x.*`
+3. Merge the functionalities of `torchx` CLI to
+   `torchrun` (a python [`console_script`](https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html)
+   that points to `torch.distributed.run`) 
+   allowing  `torchrun` to:
+     1. Submit PyTorch scripts as a job to remote schedulers (9 x OSS schedulers + 2 x FB-internal schedulers)
+     2. Run (simulate) multi-node distributed jobs locally on a single machine with multiple GPUs
+4. (alternatively) keep `torchrun` for BC (eventually to be removed), and consolidate around `torchx` CLI.
+4. (optional) Merge `torch.distributed.elastic.multiprocessing` into `torch.multiprocessing`. Adds the following features to `torch.multiprocessing`:
+     1. Launch binaries (e.g. `/home/kiuk/train.py`) in addition to functions (e.g. `pkg.my.train:main`)
+     2. Write each sub-proc's stdout and stderr to log files (configurable)
+     3. Prefix each line of stdout and stderr with local rank information
+     4. Propagate exception traces from sub-proc back to the main proc for better error handling + propagation
+
+### Backwards Compatibility (BC)
+
+#### non-BC Breaking Changes
+
+1. `torch/distributed/launch.py` and `torch/distributed/run.py` would have to remain for a few releases,
+   but will both point to `torch/x/run.py`.
```

Comment:
The consolidation makes a lot of sense to me. Do you and your plan plan to support this change all the way till `torch/distributed/launch.py` and `torch/distributed/run.py` are eventually removed from the repo?

Reply:
Yep that's the plan. Ideally we'd like to keep `torchrun` still in pytorch (which is what this RFC is advocating for) but `torch/distributed/launch.py | run.py` would be moved out to `torch.x.run`